### ISSUE TITLE
ArgBinder.bind's behaviour on invalid validators should change

### ISSUE THREAD

**Issue by BradleyChatha** [BradleyChatha/jcli#13]
While that is definitely a benefit (and quite a large one), I wonder if it's worth it on its own to change over to it, especially since the error checking for `ArgBinder` could be improved.

The main reason ArgBinder will silently ignore invalid validators is because [of a design choice](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L106) where `ArgBinder.bind` will filter out any UDAs it doesn't understand, so the calling code doesn't need to perform filtering.

These are the two checks that `ArgBinder.bind` uses:

```d
 enum canPreValidate(alias V) = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onPreValidate("", s); });
 enum canValidate(alias V)    = isStruct!V && __traits(compiles, { string s; bool a = typeof(V).init.onValidate(T.init, s); });
```

Which hides/mask certain issues, so things like [version(JCLI_BinderCompilerErrors)](https://github.com/BradleyChatha/jcli/blob/master/source/jaster/cli/binder.d#L150) exist.

But looking at it, I think the better behaviour would be to just see if the UDA has a member named `onValidate` and `onPreValidate`, and just try to call it anyway, and let the compiler dish out an error message for us.

So in other words, if the UDA has `onValidate` and/or `onPreValidate`, then recognise it as a validator regardless and just let the compiler generate an error, instead of skipping over it silently.

_Originally posted by @BradleyChatha in https://github.com/BradleyChatha/jcli/pull/5#discussion_r517644810_

**Comment by BradleyChatha** [comment]
Closed by #33
